lib/internal/test_runner/test.js

```diff
@@ -337,13 +360,13 @@ class Test extends AsyncResource {
             'passed a callback but also returned a Promise',
             kCallbackAndPromisePresent
           ));
-          await ret;
+          await testTimeout(ret, this.timeout);
         } else {
-          await promise;
+          await testTimeout(promise, this.timeout);
         }
       } else {
         // This test is synchronous or using Promises.
-        await this.runInAsyncScope(this.fn, ctx, ctx);
+        await testTimeout(this.runInAsyncScope(this.fn, ctx, ctx), this.timeout);
```

Comment:
nope

Reply:
That might be unexpected for users. Isn't the purpose of timeouts to prevent tests from getting stuck indefinitely? (I realize that the nature of JavaScript does not make it simple to have synchronous time limits, but other languages have faced the same issues.)

Reply:
interesting, how can such a thing be implemented?

Reply:
AFAIK, Jest doesn't implement this if you keep the event loop busy. Jest won't timeout because it doesn't have a chance to run timeout logic when the event loop is busy.

I think this feature could be achieved by distributing tests to a child process or worker thread, but I think it will make things over complicated.